literal/utils.go

```diff
@@ -0,0 +1,131 @@
+package literal
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/substrait-io/substrait-go/expr"
+	"github.com/substrait-io/substrait-go/types"
+)
+
+func NewBool(value bool) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[bool](value, false), nil
+}
+
+func NewInt8(value int8) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[int8](value, false), nil
+}
+
+func NewInt16(value int16) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[int16](value, false), nil
+}
+
+func NewInt32(value int32) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[int32](value, false), nil
+}
+
+func NewInt64(value int64) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[int64](value, false), nil
+}
+
+func NewFloat32(value float32) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[float32](value, false), nil
+}
+
+func NewFloat64(value float64) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[float64](value, false), nil
+}
+
+func NewString(value string) (expr.Literal, error) {
+	return expr.NewPrimitiveLiteral[string](value, false), nil
+}
+
+func NewDate(days int) (expr.Literal, error) {
+	return expr.NewLiteral[types.Date](types.Date(days), false)
+}
+
+func NewTime(tm time.Time) (expr.Literal, error) {
+	startOfTheDay := time.Date(tm.Year(), tm.Month(), tm.Day(), 0, 0, 0, 0, tm.Location())
+	return expr.NewLiteral[types.Time](types.Time(tm.Sub(startOfTheDay).Microseconds()), false)
+}
+
+func NewTimeFromMicros(micros int64) (expr.Literal, error) {
+	if micros < 0 || micros >= (24*time.Hour).Microseconds() {
+		return nil, fmt.Errorf("invalid time value %d", micros)
+	}
+	return expr.NewLiteral[types.Time](types.Time(micros), false)
+}
+
+func NewTimestamp(tm time.Time) (expr.Literal, error) {
+	return expr.NewLiteral[types.Timestamp](types.Timestamp(tm.UnixMicro()), false)
```

Comment:
this or above should probably confirm that days are only included in one and not the other. also, need better logic around things like timezone. What is expectation here if someone passes in a timezone?

if time.Time can represent many things, it seems like we need to make sure that when used to construct literals, it represents only right things and translation is reliable.

Reply:
I agree with @jacques-n here, that we need to clarify what happens if the passed in time has a location that isn't UTC. Do we ignore it and just use the UTC value? Do we handle it? etc.

Reply:
I added docs.

Time literal only represents time of the day (in microseconds)
Both Timestamp and TimestampTz use only microseconds in proto. No timezone field in proto for TimestampTz. Both APIs that I added uses microsecs since epoch of UTC. 
Timestamp (is timezone unaware)
TimestampTz (timezone aware) 

My understanding: A column with Timestamp type cannot be used with a timezone. It is a timestamp in unspecified time zone. 
A column with TimestampTz has a timestamp in micros seconds since epoch in UTC timezone. It can be localised to any timezone, as translations and comparison can be done correctly.

Reply:
Can we error on arguments that don't make sense? For example, if I pass in a time.Time that includes days to to the time literal creator.

Also, if time.Time is carrying a timezone, how is that interpereted in the timezoneless time? Converted to UTC first?

Reply:
time.Time internally stores only seconds, nanoSeconds (since 1 Jan 1885) and location. time.UnixMicro() gives micro seconds since 1 Jan 1971 UTC. This is what we need in Timestamp & TimestampTz.

Logically time.Time always stores (secs & nanosecs) in UTC. So no timezone conversion needed when getting UnixMicro().

location in Time used only to get Year(), Month(), Day(), hour(), min() 

Given the above details. I feel NewTime(time.Time) need not throw error and just take timeOfTheDay in micro seconds to create the literal. I think I captured this in the doc. May be the doc can be improved?
 or may be get rid of this API if this is going to be confusing for users? Just provide only NewTimeFromMicros(micros int64).

Reply:
The confusion I have is if I pass in a time for something like 12/30/2024 2:30pm into `NewTime`, what do I get? Would I get 2:30pm and it would silently ignore the date? Would it store the date? From my pov, I would expect to error and say something like "time not within allowed range". Looking at the spec, it is listed at the following valid timerange: 0..86,399,999,999 microseconds. Outside that should error.

Reply:
We do range check in NewTimeFromMicros.

NewTime(time.Time) current code takes 2.30 pm and ignores the date part. My point is if we throw error in this case, this api is not so useful. For some one to use this api, they have to create a time like "1 Jan 1971 2:30PM" and pass it. Should we get rid of this API?

May be a better api is NewTime(hours, minutes, seconds, microsecs int32).

Reply:
yeah, that sounds like better api.

we should range check it. someone can't say 25 hours, for example.